## Exercises: Explore the dataset

In [58]:
import pandas as pd
import seaborn as sns
taxis = sns.load_dataset("taxis")

**Explore the "taxis" dataset to answer the following questions:**

In [59]:
taxis.head()

pickup             dropoff  passengers  distance  fare   tip  \
0 2019-03-23 20:21:09 2019-03-23 20:27:24           1      1.60   7.0  2.15   
1 2019-03-04 16:11:55 2019-03-04 16:19:00           1      0.79   5.0  0.00   
2 2019-03-27 17:53:01 2019-03-27 18:00:25           1      1.37   7.5  2.36   
3 2019-03-10 01:23:59 2019-03-10 01:49:51           1      7.70  27.0  6.15   
4 2019-03-30 13:27:42 2019-03-30 13:37:14           3      2.16   9.0  1.10   

   tolls  total   color      payment            pickup_zone  \
0    0.0  12.95  yellow  credit card        Lenox Hill West   
1    0.0   9.30  yellow         cash  Upper West Side South   
2    0.0  14.16  yellow  credit card          Alphabet City   
3    0.0  36.95  yellow  credit card              Hudson Sq   
4    0.0  13.40  yellow  credit card           Midtown East   

            dropoff_zone pickup_borough dropoff_borough  
0    UN/Turtle Bay South      Manhattan       Manhattan  
1  Upper West Side South      Manhattan       Manhattan  
2           West Village      Manhattan       Manhattan  
3         Yorkville West      Manhattan       Manhattan  
4         Yorkville West      Manhattan       Manhattan

**Q1:** How many rows and column are in the dataset?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Rows:</b> 6433
&nbsp;&nbsp;&nbsp;<b>Columns:</b> 14
</details>

In [60]:
rows, columns =  taxis.shape
print(f"Dataframe has {rows} rows, and {columns} columns.")

rows = len(taxis)
columns = len(taxis.columns)
print(f"Dataframe has {rows} rows, and {columns} columns.")

Dataframe has 6433 rows, and 14 columns.
Dataframe has 6433 rows, and 14 columns.


**Q2:** What datatype is the most common in the set?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;object (6 columns)
</details>

In [61]:
taxis.info()

taxis.dtypes.value_counts()    # Lista med kolumner och vad varje en har för datatyp, med value counts så summerar man sedan alla datatyper. 
                               # Vill man istället ha summa på allt får man lägga till .sum() på slutet


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6433 entries, 0 to 6432
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   pickup           6433 non-null   datetime64[ns]
 1   dropoff          6433 non-null   datetime64[ns]
 2   passengers       6433 non-null   int64         
 3   distance         6433 non-null   float64       
 4   fare             6433 non-null   float64       
 5   tip              6433 non-null   float64       
 6   tolls            6433 non-null   float64       
 7   total            6433 non-null   float64       
 8   color            6433 non-null   object        
 9   payment          6389 non-null   object        
 10  pickup_zone      6407 non-null   object        
 11  dropoff_zone     6388 non-null   object        
 12  pickup_borough   6407 non-null   object        
 13  dropoff_borough  6388 non-null   object        
dtypes: datetime64[ns](2), float64(5), int64(

object            6
float64           5
datetime64[ns]    2
int64             1
Name: count, dtype: int64

**Q3:** What is the average number of passengers in a taxi?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;1.54
</details>

In [62]:
taxis["passengers"].mean().round(2)

1.54

**Q4:** What is the most common number of passengers in a taxi?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;1
</details>

In [63]:
taxis["passengers"].mode()

0    1
Name: passengers, dtype: int64

**Q5:** What is the most common payment method?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;credit card
</details>

In [64]:
taxis["payment"].mode()

0    credit card
Name: payment, dtype: object

**Q6:** Which of the categorical features has the most categories?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;dropoff_zone (203 categories)
</details>

In [65]:
# taxis.columns      # Just to get list of columns and use belo
cat_columns = taxis[['color', 'payment', 'pickup_zone', 'dropoff_zone',   # Create column
       'pickup_borough', 'dropoff_borough']]


for col in cat_columns:                                 # By a for loop getting sum of unique catogries
    print(f"{col}: {taxis[col].nunique()}")





color: 2
payment: 2
pickup_zone: 194
dropoff_zone: 203
pickup_borough: 4
dropoff_borough: 5


**Q7:** What percentage of cars in the set are yellow?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;84.7%
</details>

In [66]:
total_taxis = len(taxis) 
total_yellow_taxis = len(taxis[taxis["color"] == "yellow"])

percentage = total_yellow_taxis/total_taxis * 100

print(f"The percentage of yellow taxis are {percentage:.1F} %")

The percentage of yellow taxis are 84.7 %


**Q8:** Which dropoff borough is most common? Which one is least common?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Most common:</b> Manhattan (5206)<br>
&nbsp;&nbsp;&nbsp;<b>Least common:</b> Staten Island (2)<br>
</details>

In [67]:
display(taxis["dropoff_borough"].mode())

display(taxis["dropoff_borough"].value_counts())

0    Manhattan
Name: dropoff_borough, dtype: object

dropoff_borough
Manhattan        5206
Queens            542
Brooklyn          501
Bronx             137
Staten Island       2
Name: count, dtype: int64

**Q9:** Which column has the most missing values? How many?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;<i>dropoff_zone</i> and <i>dropoff_borough</i> both have 45 missing values.
</details>

In [68]:
taxis.isna().sum()

pickup              0
dropoff             0
passengers          0
distance            0
fare                0
tip                 0
tolls               0
total               0
color               0
payment            44
pickup_zone        26
dropoff_zone       45
pickup_borough     26
dropoff_borough    45
dtype: int64

### Memory usage
``` taxis.info(memory_usage="deep") ``` gives you the total memory usage of the dataframe.

``` taxis.memory_usage(deep=True) ``` give you the total memory usage for each column.

**Answer the following questions:**

**Q10:** What is the total memory usage of the dataframe?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;2.9 MB
</details>

In [69]:
taxis.info(memory_usage="deep")


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6433 entries, 0 to 6432
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   pickup           6433 non-null   datetime64[ns]
 1   dropoff          6433 non-null   datetime64[ns]
 2   passengers       6433 non-null   int64         
 3   distance         6433 non-null   float64       
 4   fare             6433 non-null   float64       
 5   tip              6433 non-null   float64       
 6   tolls            6433 non-null   float64       
 7   total            6433 non-null   float64       
 8   color            6433 non-null   object        
 9   payment          6389 non-null   object        
 10  pickup_zone      6407 non-null   object        
 11  dropoff_zone     6388 non-null   object        
 12  pickup_borough   6407 non-null   object        
 13  dropoff_borough  6388 non-null   object        
dtypes: datetime64[ns](2), float64(5), int64(

**Q11:** Which column takes up the most memory? How many kilobytes?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;pickup_zone (470 KB)
</details>

In [70]:
taxis.memory_usage(deep=True)

Index                 132
pickup              51464
dropoff             51464
passengers          51464
distance            51464
fare                51464
tip                 51464
tolls               51464
total               51464
color              404297
payment            423176
pickup_zone        469744
dropoff_zone       469466
pickup_borough     420944
dropoff_borough    420381
dtype: int64

**Q12:** Why does the numeric columns all take up exactly 51464 bytes?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;They all use 64 bit datatypes. 64 bits = 8 bytes. 6433 entries * 8 bytes = 51464 bytes.
</details>

**Q13:** What is the total memory usage after converting all *object* columns to *category*?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;494.0 KB
</details>

In [75]:
taxis.dtypes

pickup             datetime64[ns]
dropoff            datetime64[ns]
passengers                  int64
distance                  float64
fare                      float64
tip                       float64
tolls                     float64
total                     float64
color                      object
payment                    object
pickup_zone                object
dropoff_zone               object
pickup_borough             object
dropoff_borough            object
dtype: object

In [79]:
taxis = taxis.astype({"color": 'category', "payment":'category', "pickup_zone": 'category', "dropoff_zone":'category', "pickup_borough":'category',
"dropoff_borough":'category'})

taxis.dtypes
taxis.info(memory_usage="deep")


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6433 entries, 0 to 6432
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   pickup           6433 non-null   datetime64[ns]
 1   dropoff          6433 non-null   datetime64[ns]
 2   passengers       6433 non-null   int64         
 3   distance         6433 non-null   float64       
 4   fare             6433 non-null   float64       
 5   tip              6433 non-null   float64       
 6   tolls            6433 non-null   float64       
 7   total            6433 non-null   float64       
 8   color            6433 non-null   category      
 9   payment          6389 non-null   category      
 10  pickup_zone      6407 non-null   category      
 11  dropoff_zone     6388 non-null   category      
 12  pickup_borough   6407 non-null   category      
 13  dropoff_borough  6388 non-null   category      
dtypes: category(6), datetime64[ns](2), float

**Q14:** ... and after also converting *float64* to *float32*?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;368.4 KB
</details>

In [81]:
taxis = taxis.astype({"distance": 'float32', "fare": 'float32', "tip": 'float32', "tolls":'category', "total": 'float32'})
taxis.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6433 entries, 0 to 6432
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   pickup           6433 non-null   datetime64[ns]
 1   dropoff          6433 non-null   datetime64[ns]
 2   passengers       6433 non-null   int64         
 3   distance         6433 non-null   float32       
 4   fare             6433 non-null   float32       
 5   tip              6433 non-null   float32       
 6   tolls            6433 non-null   category      
 7   total            6433 non-null   float32       
 8   color            6433 non-null   category      
 9   payment          6389 non-null   category      
 10  pickup_zone      6407 non-null   category      
 11  dropoff_zone     6388 non-null   category      
 12  pickup_borough   6407 non-null   category      
 13  dropoff_borough  6388 non-null   category      
dtypes: category(7), datetime64[ns](2), float

**Q15:** What is the smallest datatype we can convert passengers to? What is the total memory usage after converting passengers to the new type?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;The maximum number of passengers in the dataset are 6,<br> 
&nbsp;&nbsp;&nbsp;and therefore the values easily fit into the <i>int8</i> type (8 bit integer).<br>
<br>
&nbsp;&nbsp;&nbsp;New size: 324.4 KB
</details>

**Q16:** How many percent of the orignal datasize is the new dataset after converting all the types as above?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;11.0 %
</details>

### Final note:
Just to be clear, if we want to limit our memory usage by specifying datatypes with a smaller memory footprint, it makes more sense to do so when loading the dataset in to pandas, than changing the type afterwards (as in the example above).

Most common ways to load data into pandas (like pd.from_csv, pd.from_json etc) provides optional parameters for setting the datatype as the files are read into pandas dataframes.

Also, note that this is really only a concern when working with huge sets of data. For smaller datasets, like the one in the example above, it doesn't really matter, and might be only unneccessary work to optimize. The above exercises just serve as examples to better understand data types and their memory footprints.